org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/outline/ModelBuilder.scala

```diff
@@ -0,0 +1,417 @@
+          else
+            VarNode(name.decodedName.toString(), parent, if (tpt.isEmpty) None else Some(showType(tpt)))
+          setPos(ch, t)
+          ch.setFlags(mods.flags)
+          parent.addChild(ch)
+        }
+
+        case DefDef(mods, name, tparamss, vparamss, tpt, rsh) =>
+          def typeList = {
+            vparamss.map { x =>
+              {
+                if (!x.isEmpty) {
+                  val h = x.head
+                  val prefix: String = if ((h.mods.flags & IMPLICIT) != 0) "implicit " else ""
+                  prefix + h.name + ": " + showType(h.tpt) :: x.tail.map(v => v.name + ": " + showType(v.tpt))
+                } else
+                  List()
+              }
+            }
+
+          }
+          val ch = MethodNode(name.decodedName.toString(), parent, typeList)
+          ch.returnType = if (!tpt.isEmpty) Some(showType(tpt)) else None
+          setPos(ch, t)
+          ch.setFlags(mods.flags)
+          parent.addChild(ch)
+          updateTree(ch, rsh)
+
+        case ModuleDef(mods, name, _) => {
+          val ch = ObjectNode(name.decodedName.toString(), parent)
+          setPos(ch, t)
+          ch.setFlags(mods.flags)
+          parent.addChild(ch)
+          t.children.foreach(x => updateTree(ch, x))
+        }
+
+        case DocDef(comment: DocComment, definition: Tree) => updateTree(parent, definition)
+
+        case Block(stats: List[Tree], expr: Tree) => {
+          stats.foreach(updateTree(parent, _))
+        }
+
+        case TypeDef(mods: Modifiers, name: TypeName, tparams: List[TypeDef], rhs: Tree) =>
+          val ch = TypeNode(name.decodedName.toString(), parent)
+          setPos(ch, t)
+          ch.setFlags(mods.flags)
+          if (!ch.isParam)
+            parent.addChild(ch)
+
+        case Import(expr: Tree, selectors) =>
+          def printImport = {
+            val sb = new StringBuilder
+            sb.append(expr.toString())
+            if (selectors.size == 1)
+              sb.append("." + selectors.head.name)
+            else {
+              sb.append(".{")
+              selectors.foreach(s => sb.append(s.name + ","))
+              sb.setLength(sb.length - 1)
+              sb.append("}")
+            }
+            sb.toString
+          }
+          parent.last match {
+            case Some(p: ImportsNode) => {
+              val in = ImportNode(printImport, p)
+              p.addChild(in)
+              setPos(in, t)
+            }
+            case _ => {
+              val ip = ImportsNode(parent)
+              parent.addChild(ip)
+              val in = ImportNode(printImport, ip)
+              ip.addChild(in)
+              setPos(in, t)
+            }
+          }
+
+        case _ => //logger.info("_"+t.getClass+", p="+ t.toString)
+      }
+    }
+    val rootNode = new RootNode
+    val t = comp.parseTree(src)
+    comp.asyncExec(updateTree(rootNode, t)).get
```

Comment:
Is this call needed? Do you call any methods on `Symbol` or `Type` inside? If not, there shouldn't be a need to execute this on the compiler thread. Parse trees are the only things that are thread-safe in the compiler. :)